src/Datadog.Trace/DatadogSpanStagingArea.cs

```diff
@@ -0,0 +1,130 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace
+{
+    internal static class DatadogSpanStagingArea
+    {
+        private static readonly ILog Log = LogProvider.GetLogger(typeof(DatadogSpanStagingArea));
+        private static readonly ConcurrentDictionary<Guid, Action> WakeUpTasks = new ConcurrentDictionary<Guid, Action>();
+        private static readonly ConcurrentQueue<FlushTask> FlushTaskQueue = new ConcurrentQueue<FlushTask>();
+
+        public static int FlushTaskCount { get; private set; }
+
+        public static DateTime LastFlushRequest { get; private set; }
+
+        public static TimeSpan TimeSinceLastFlush => Now.Subtract(LastFlushRequest);
+
+        private static DateTime Now => DateTime.UtcNow;
+
+        public static void RegisterForWakeup(Action wakeupCall)
+        {
+            WakeUpTasks.TryAdd(Guid.NewGuid(), wakeupCall);
+        }
+
+        public static void QueueSpanForFlush(Span span)
```

Comment:
This needs to be re-worked a little bit. 
It's fine to queue spans as they finish, but it's not fine to flush spans before the root span is in.
The agent currently assumes that the first span in a trace submission is the root span, so until that is re-worked, we need to submit the parent first. 
There are other considerations IIRC as well, like figuring out counts/metrics, as the agent assumes that the whole trace will be submitted in one request, which obviously doesn't work out for asynchronous operations that will run after the root span closes.
To be continued.

Reply:
@brettlangdon Did we discover this isn't the case now?